## Exploring Ebay Kleinanzeigen Car Sales Data
based on 2016 data on used cars from Ebay Kleinanzeigen.

The data is 370000 used cars scraped with Scrapy from Ebay-Kleinanzeigen. 
The aim of this project will be to clean and analyze the dataset using numpy and pandas.
The content of the data is in german, so the fields had to be translated and the following fields are included:

    dateCrawled : when this ad was first crawled, all field-values are taken from this date
    name : "name" of the car
    seller : private or dealer
    offerType
    price : the price on the ad to sell the car
    abtest
    vehicleType
    yearOfRegistration : at which year the car was first registered
    gearbox
    powerPS : power of the car in PS
    model
    kilometer : how many kilometers the car has driven
    monthOfRegistration : at which month the car was first registered
    fuelType
    brand
    notRepairedDamage : if the car has a damage which is not repaired yet
    dateCreated : the date for which the ad at ebay was created
    nrOfPictures : number of pictures in the ad
    postalCode
    lastSeenOnline : when the crawler saw this ad last online

The original dataset is available at https://data.world/data-society/used-cars-data.

In [49]:
import pandas as pd
import numpy as np


Using .info() and .head(), I can visualize information about the dataset, as well as view the desired number of columns.

In [50]:
autos = pd.read_csv('autos.csv', encoding='Windows-1252')
autos.info()
autos.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

dateCrawled                                               name  \
0  2016-03-24 11:52:17                                         Golf_3_1.6   
1  2016-03-24 10:58:45                               A5_Sportback_2.7_Tdi   
2  2016-03-14 12:52:21                     Jeep_Grand_Cherokee_"Overland"   
3  2016-03-17 16:54:04                                 GOLF_4_1_4__3TÜRER   
4  2016-03-31 17:25:20                     Skoda_Fabia_1.4_TDI_PD_Classic   
5  2016-04-04 17:36:23  BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...   
6  2016-04-01 20:48:51                        Peugeot_206_CC_110_Platinum   
7  2016-03-21 18:54:38                       VW_Derby_Bj_80__Scheunenfund   
8  2016-04-04 23:42:13               Ford_C___Max_Titanium_1_0_L_EcoBoost   
9  2016-03-17 10:53:50  VW_Golf_4_5_tuerig_zu_verkaufen_mit_Anhaengerk...   

   seller offerType  price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot    480     test         NaN                1993   
1  privat   Angebot  18300     test       coupe                2011   
2  privat   Angebot   9800     test         suv                2004   
3  privat   Angebot   1500     test  kleinwagen                2001   
4  privat   Angebot   3600     test  kleinwagen                2008   
5  privat   Angebot    650     test   limousine                1995   
6  privat   Angebot   2200     test      cabrio                2004   
7  privat   Angebot      0     test   limousine                1980   
8  privat   Angebot  14500  control         bus                2014   
9  privat   Angebot    999     test  kleinwagen                1998   

     gearbox  powerPS    model  kilometer  monthOfRegistration fuelType  \
0    manuell        0     golf     150000                    0   benzin   
1    manuell      190      NaN     125000                    5   diesel   
2  automatik      163    grand     125000                    8   diesel   
3    manuell       75     golf     150000                    6   benzin   
4    manuell       69    fabia      90000                    7   diesel   
5    manuell      102      3er     150000                   10   benzin   
6    manuell      109  2_reihe     150000                    8   benzin   
7    manuell       50   andere      40000                    7   benzin   
8    manuell      125    c_max      30000                    8   benzin   
9    manuell      101     golf     150000                    0      NaN   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0  volkswagen               NaN  2016-03-24 00:00:00             0   
1        audi                ja  2016-03-24 00:00:00             0   
2        jeep               NaN  2016-03-14 00:00:00             0   
3  volkswagen              nein  2016-03-17 00:00:00             0   
4       skoda              nein  2016-03-31 00:00:00             0   
5         bmw                ja  2016-04-04 00:00:00             0   
6     peugeot              nein  2016-04-01 00:00:00             0   
7  volkswagen              nein  2016-03-21 00:00:00             0   
8        ford               NaN  2016-04-04 00:00:00             0   
9  volkswagen               NaN  2016-03-17 00:00:00             0   

   postalCode             lastSeen  
0       70435  2016-04-07 03:16:57  
1       66954  2016-04-07 01:46:50  
2       90480  2016-04-05 12:47:46  
3       91074  2016-03-17 17:40:17  
4       60437  2016-04-06 10:17:21  
5       33775  2016-04-06 19:17:07  
6       67112  2016-04-05 18:18:39  
7       19348  2016-03-25 16:47:58  
8       94505  2016-04-04 23:42:13  
9       27472  2016-03-31 17:17:06

In [51]:
autos = autos.replace ([np.inf, -np.inf], np.nan)

## Data Exploration and Cleaning

When looking at the columns, I notice that they are in camelcase, and that some of the names are a little bit unclear, therefore I use the .columns function to change the column names.

In [52]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [53]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price ab_test vehicle_type  registration_year    gearbox  power_ps  model  \
0    480    test          NaN               1993    manuell         0   golf   
1  18300    test        coupe               2011    manuell       190    NaN   
2   9800    test          suv               2004  automatik       163  grand   
3   1500    test   kleinwagen               2001    manuell        75   golf   
4   3600    test   kleinwagen               2008    manuell        69  fabia   

   odometer  registration_month fuel_type       brand unrepaired_damage  \
0    150000                   0    benzin  volkswagen               NaN   
1    125000                   5    diesel        audi                ja   
2    125000                   8    diesel        jeep               NaN   
3    150000                   6    benzin  volkswagen              nein   
4     90000                   7    diesel       skoda              nein   

            ad_created  num_photos  postal_code            last_seen  
0  2016-03-24 00:00:00           0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00           0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00           0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00           0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00           0        60437  2016-04-06 10:17:21

First, I'll explore the data to find obvious areas of where cleaning needs to happen for further analysis.

In [54]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

       ab_test vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model       odometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  num_photos   postal_code  \
count             299468               371528    371528.0  371528.00000   
unique                 2                  114         NaN           NaN   
top                 nein  2016-04-03 00:00:00         NaN           NaN   
freq              263182                14450         NaN           NaN   
mean                 NaN                  NaN         0.0   50820.66764   
std                  NaN                  NaN         0.0   25799.08247   
min                  NaN                  NaN         0.0    1067.00000   
25%                  NaN                  NaN         0.0   30459.00000   
50%                  NaN                  NaN         0.0   49610.00000   
75%                  NaN                  NaN         0.0   71546.00000   
max                  NaN                  NaN         0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

I can see that many of the columns show "NaN" or "Not a Number". 
However, one of the values that doesn't make sense is the num_photos column that shows 0.0. I will investigate this further.

In [55]:
autos["num_photos"].value_counts()

0    371528
Name: num_photos, dtype: int64

It is obvious that the num_photos column has 0 for every column. I will get rid of this column, or "drop" it, and will do the same for the "seller" and "offer_type" columns as they also mostly have one value. 

In [56]:
autos = autos.drop(["num_photos", "seller", "offer_type"], axis=1)



Next, I will remove the extra characters being stored as text in the price and auto columns, and then convert those columns to the integer type. 

In [57]:
autos['price'] = autos['price'].apply(str).str.replace('$', '')
autos['price'] = autos['price'].apply(str).str.replace(",","")
autos['price'] = autos['price'].astype(int)
autos["price"].head()

0      480
1    18300
2     9800
3     1500
4     3600
Name: price, dtype: int32

And, I will do the same, removing all non-numeric characters for the "odometer" column, as well as convert all the values to the integer type. I will also rename the column to "odometer_km" to make it clearer of the unit we are working with.

In [58]:
autos['odometer'] = autos['odometer'].apply(str).str.replace('km', '')
autos['odometer'] = autos['odometer'].apply(str).str.replace(",","")
autos['price'] = autos['price'].astype(int)
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)
autos["price"].head()

0      480
1    18300
2     9800
3     1500
4     3600
Name: price, dtype: int32

## Exploring the Odometer and Price columns

In [59]:
autos["odometer_km"].value_counts()

150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: odometer_km, dtype: int64

I can see that the values in this field are rounded, which shows us that sellers probably had to choose from pre-set options for this field. It is also very clear that there are more high mileage than low mileage vehicles. 

In [60]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
autos["price"].value_counts().head(20)

(5597,)
count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64


0       10778
500      5670
1500     5394
1000     4649
1200     4594
2500     4438
600      3819
3500     3792
800      3784
2000     3432
999      3364
750      3203
650      3150
4500     3053
850      2946
2200     2936
700      2936
1800     2886
900      2874
950      2793
Name: price, dtype: int64

For prices, it does seem like at least the top 20 most frequent prices are round numbers; however since there are 5597 prices I'm sure that is not the case for the entire dataset. The first thing I can notice is that the most common price is 0 dollars, with 10778 cars, and therefore I might want to delete these cars. However, first, let's look at the highest prices.

In [61]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
Name: price, dtype: int64

In [62]:
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
11        5
12        8
13        7
14        5
15       27
16        2
17        5
18        3
19        3
20       51
Name: price, dtype: int64

Now I can see that the most expensive car is listed at a price of 2 billion euros, which I definitely need to get rid of, as well as a few of the other higher prices which are definitely going to skew the data the wrong way. I will choose 350 000 as the max value and 1 as the minimum value for "price" as I think that a car sold on eBay should probably not cost more than 350 000, and some items could legitimately have been set at 1 euro because the opening bid could have been 1 euro. 

In [63]:
autos = autos[autos["price"].between(1,351000)]
autos["price"].describe()

count    360635.000000
mean       5898.671956
std        8866.359669
min           1.000000
25%        1250.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

The mean has now decreased to 5990 euros, and the max value is 350 000, which shows me that the last command worked. I now have a much more accurate dataset.

## Exploring the date columns

There are a few columns with date information: date_crawled, registration_month, registration_year, ad_created, and last_seen. Let's explore some of these to get a better idea of the dataset. 

In [64]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57
1  2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21

I will look at the distribution of values in the date_crawled, ad_created, and last_seen columns but I will view these as percentages instead of counts by chaining the series.value_counts(normalize=True, dropna=False) method.

In [65]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.025547
2016-03-06    0.014483
2016-03-07    0.035657
2016-03-08    0.033469
2016-03-09    0.034115
2016-03-10    0.032645
2016-03-11    0.032773
2016-03-12    0.036242
2016-03-13    0.015783
2016-03-14    0.036330
2016-03-15    0.033424
2016-03-16    0.030205
2016-03-17    0.031647
2016-03-18    0.013119
2016-03-19    0.035271
2016-03-20    0.036400
2016-03-21    0.035682
2016-03-22    0.032493
2016-03-23    0.032002
2016-03-24    0.029914
2016-03-25    0.032800
2016-03-26    0.031974
2016-03-27    0.030227
2016-03-28    0.035063
2016-03-29    0.034126
2016-03-30    0.033535
2016-03-31    0.031872
2016-04-01    0.034145
2016-04-02    0.035094
2016-04-03    0.038812
2016-04-04    0.037628
2016-04-05    0.012780
2016-04-06    0.003128
2016-04-07    0.001617
Name: date_crawled, dtype: float64

In [66]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_values()
        )

2016-04-07    0.001617
2016-04-06    0.003128
2016-04-05    0.012780
2016-03-18    0.013119
2016-03-06    0.014483
2016-03-13    0.015783
2016-03-05    0.025547
2016-03-24    0.029914
2016-03-16    0.030205
2016-03-27    0.030227
2016-03-17    0.031647
2016-03-31    0.031872
2016-03-26    0.031974
2016-03-23    0.032002
2016-03-22    0.032493
2016-03-10    0.032645
2016-03-11    0.032773
2016-03-25    0.032800
2016-03-15    0.033424
2016-03-08    0.033469
2016-03-30    0.033535
2016-03-09    0.034115
2016-03-29    0.034126
2016-04-01    0.034145
2016-03-28    0.035063
2016-04-02    0.035094
2016-03-19    0.035271
2016-03-07    0.035657
2016-03-21    0.035682
2016-03-12    0.036242
2016-03-14    0.036330
2016-03-20    0.036400
2016-04-04    0.037628
2016-04-03    0.038812
Name: date_crawled, dtype: float64

In [67]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001264
2016-03-06    0.004098
2016-03-07    0.005202
2016-03-08    0.007939
2016-03-09    0.009824
2016-03-10    0.011460
2016-03-11    0.012955
2016-03-12    0.023240
2016-03-13    0.008410
2016-03-14    0.012176
2016-03-15    0.016324
2016-03-16    0.016418
2016-03-17    0.028699
2016-03-18    0.006888
2016-03-19    0.016330
2016-03-20    0.019884
2016-03-21    0.020026
2016-03-22    0.020508
2016-03-23    0.018015
2016-03-24    0.019163
2016-03-25    0.019000
2016-03-26    0.015958
2016-03-27    0.016721
2016-03-28    0.022189
2016-03-29    0.023284
2016-03-30    0.023725
2016-03-31    0.024243
2016-04-01    0.023897
2016-04-02    0.024967
2016-04-03    0.025308
2016-04-04    0.025536
2016-04-05    0.126962
2016-04-06    0.218950
2016-04-07    0.130437
Name: last_seen, dtype: float64

In [68]:
print(autos["ad_created"].str[:10].unique().shape)
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

(114,)


2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
                ...   
2016-04-03    0.039001
2016-04-04    0.037736
2016-04-05    0.011613
2016-04-06    0.003119
2016-04-07    0.001553
Name: ad_created, Length: 114, dtype: float64

And, I use the series.describe() function to understand the distribution of registration_year.

In [69]:
autos["registration_year"].describe()

count    360635.000000
mean       2004.433133
std          81.016977
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

I can see that oddly, the minimum registration year is the year 1000, and the maximum is 9999, which is obviously some sort of error, as no cars from those years would be on sale. Therefore, I use the following line of code to figure out what percentage of cars in the data lie outside of the registration years 1900-2016.

In [70]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0] * 100

3.8751091824143526

I find that 3.8 percent of cars lie outside of 1900-2016. The following line of code will remove these from the dataset.

In [71]:
autos = autos[autos["registration_year"].between(1900,2016)]
print (autos["registration_year"].describe())
autos["registration_year"].value_counts(normalize=True).head(10) * 100

count    346660.000000
mean       2002.896489
std           7.244719
min        1910.000000
25%        1999.000000
50%        2003.000000
75%        2008.000000
max        2016.000000
Name: registration_year, dtype: float64


2000    6.669936
1999    6.355218
2005    6.266947
2006    5.770784
2001    5.695494
2003    5.655686
2004    5.617320
2002    5.428951
2007    5.049905
1998    4.969134
Name: registration_year, dtype: float64

I think that now, my dataset is cleaned enough and I can move onto to further analysis. 

## Aggregation of the brand column

Next, I will use an analysis technique that I recently learned called aggregation to explore varioations across different car brands. First, I print the distribution across all car brands to check which brands are the most represented.

In [72]:
autos["brand"].value_counts(normalize=True) * 100

volkswagen        21.170023
bmw               10.987134
opel              10.640974
mercedes_benz      9.684129
audi               8.954307
ford               6.891767
renault            4.751630
peugeot            3.015346
fiat               2.569088
seat               1.866094
skoda              1.568684
mazda              1.538395
smart              1.433104
citroen            1.395027
nissan             1.359834
toyota             1.293198
hyundai            0.997231
sonstige_autos     0.949345
mini               0.938383
volvo              0.914729
mitsubishi         0.823574
honda              0.753188
kia                0.691456
suzuki             0.636358
alfa_romeo         0.630878
porsche            0.621070
chevrolet          0.502221
chrysler           0.386257
dacia              0.249524
jeep               0.219235
land_rover         0.216639
daihatsu           0.216062
subaru             0.211735
jaguar             0.173369
saab               0.146541
daewoo             0

Now, I will select the brands that have more than 5 percent of the distribution, and add them into a list called common_brands. Ater that, I will iterate over these brands, and assign the mean price to each, in a dictionary, assigning the name as the key and the price as the value. 

In [73]:
brand_counts = autos["brand"].value_counts(normalize=True) * 100
common_brands = brand_counts[brand_counts > 5].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [74]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

print (brand_mean_prices)

{'volkswagen': 5400, 'bmw': 8449, 'opel': 2971, 'mercedes_benz': 8551, 'audi': 9086, 'ford': 3696}


The top 6 brands are VW, BMW, Opel, Mercedes, Audi, and Ford; however, the highest mean price is for Audis, closely followed by Mercedes and BMW. VW is somewhere in the middle, and Opel and Ford round out the top 6. 

Now, I'll use aggregation to understand and check the relationship between the average mileage for those brands and their mean price. To do so, I combine the data from both series objects into a single dataframe using the pandas series constructor and dataframe constructor.

First, I create a series constructor called bmp_series that uses the brand_mean_prices dictionary that I created above.

In [75]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           5400
bmw                  8449
opel                 2971
mercedes_benz        8551
audi                 9086
ford                 3696

Then, I use the following loop to calculate the mean mileage and mean price for the top 6 brands, and store these in a dictionary called brand_mean_mileage.

In [76]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = float(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

Now I am ready to create my dataframe from the first series object, and then, I can assign the other series as a new column in this dataframe. Unfortunately, the mean_mileage values show up as infinity because there must be infinity values in my dataset. I will try to figure out a way to get rid of these infinity values in the future. 
I added the line autos = autos.replace ([np.inf, -np.inf], np.nan) at the start of my code but it did not do the trick and everything else I tried didn't allow my code to run.

In [77]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
ford                    inf
audi                    inf
mercedes_benz           inf
opel                    inf
bmw                     inf
volkswagen              inf

I would then expect to have the mean mileage calculated, and they would appear here next to the mean price, and would be helpful in helping someone decide what brand of car they should look at based on those two factors.

In [78]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
ford                    inf        3696
audi                    inf        9086
mercedes_benz           inf        8551
opel                    inf        2971
bmw                     inf        8449
volkswagen              inf        5400